pkg/util/metric/metric.go

```diff
@@ -176,10 +186,261 @@ func maybeTick(m periodic) {
 	}
 }
 
+// A HdrHistogram collects observed values by keeping bucketed counts. For
+// convenience, internally two sets of buckets are kept: A cumulative set (i.e.
+// data is never evicted) and a windowed set (which keeps only recently
+// collected samples).
+//
+// Top-level methods generally apply to the cumulative buckets; the windowed
+// variant is exposed through the Windowed method.
+type HdrHistogram struct {
+	Metadata
+	maxVal int64
+	mu     struct {
+		syncutil.Mutex
+		cumulative *hdrhistogram.Histogram
+		*tickHelper
+		sliding *hdrhistogram.WindowedHistogram
+	}
+}
+
+func (h *HdrHistogram) ValueAtQuantileWindowed(q float64) float64 {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	return ValueAtQuantileWindowed(h.toPrometheusMetricWindowedLocked().Histogram, q)
+}
+
+func (h *HdrHistogram) Mean() float64 {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	return h.mu.cumulative.Mean()
+}
+
+func (h *HdrHistogram) TotalSum() float64 {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	return h.ToPrometheusMetric().GetSummary().GetSampleSum()
+}
+
+// useHdrHistogramsEnvVar can be used to switch all histograms to use the
+// legacy HDR histograms (except for those that explicitly force the use
+// of the newer Prometheus via HistogramOptions.ForceUsePrometheus).
+// HDR Histograms dynamically generate bucket boundaries, which can lead
+// to hundreds of buckets. This can cause performance issues with timeseries
+// databases like Prometheus.
+const useHdrHistogramsEnvVar = "COCKROACH_ENABLE_HDR_HISTOGRAMS"
+
+var hdrEnabled = util.ConstantWithMetamorphicTestBool(useHdrHistogramsEnvVar, envutil.EnvOrDefaultBool(useHdrHistogramsEnvVar, false))
+
+// HdrEnabled returns whether or not the HdrHistogram model is enabled
+// in the metric package. Primarily useful in tests where we want to validate
+// different outputs depending on whether or not HDR is enabled.
+func HdrEnabled() bool {
+	return hdrEnabled
+}
+
+type HistogramOptions struct {
+	Metadata           Metadata
+	Duration           time.Duration
+	MaxVal             int64
+	SigFigs            int
+	Buckets            []float64
+	ForceUsePrometheus bool
+	UseHdrLatency      bool
```

Comment:
Field comments would be helpful, as some of the options are mutually exclusive.

I think you'd really want an enum here,

```go
type HistogramMode byte

const (
  HistogramModePrometheus HistogramMode = iota // corresponds to ForceUsePrometheus
  HistogramModePreferHdr // !ForceUsePrometheus && !UseHdrLatency && hdrEnabled
  HistogramModePreferHdrLatency // ...
)
``` 

but I'm fine w/ cutting that corner.

Reply:
Added field comments, thanks for the reminder!

I also went ahead and added the enum, which I think is a great idea. However, I didn't introduce `HistogramModePreferHdr` as this is communicated via the env var, and I don't think we should ever *prefer* the HDR histograms otherwise. 

Finally, I started at `iota + 1` to avoid mistaking the zero-value with an intentionally set `HistogramMode`.